lib/bq_stream/bq_stream.rb

```diff
@@ -14,86 +14,188 @@ module BqStream
+    end
 
-  def self.create_bq_dataset
-    @bq_writer.create_dataset(dataset)
-  end
+    def big_query_check(klass, dataset, table, qty)
+      create_bq_writer
+      bq_gather_and_compare(klass, dataset, table, qty)
+    end
+
+    def database_check(klass, date_attr, dataset, table, qty = nil, start_date = Time.current.beginning_of_week - 1.week, end_date =  Time.current.end_of_week - 1.week)
+      create_bq_writer
+      query_db_records(klass.classify.constantize, date_attr, qty, start_date, end_date)
+      collect_bq_records(klass, dataset, table, @db_ids)
+      bq_gather_and_compare(klass, dataset, table, qty)
+    end
+
+    protected
 
-  def self.create_bq_table
-    bq_table_schema = { table_name:   { type: 'STRING', mode: 'REQUIRED' },
-                        record_id:    { type: 'INTEGER', mode: 'REQUIRED' },
-                        attr:         { type: 'STRING', mode: 'NULLABLE' },
-                        new_value:    { type: 'STRING', mode: 'NULLABLE' },
-                        updated_at:   { type: 'TIMESTAMP', mode: 'REQUIRED' } }
-    @bq_writer.create_table(bq_table_name, bq_table_schema)
+    def bq_gather_and_compare(klass, dataset, table, qty)
+      query_bq_records(qty, dataset, table)
+      compare_(@records, klass.classify.constantize)
+      @results.each { |result| puts result }
+    end
+
+    def query_bq_records(qty, dataset, table)
+      @schema = []
+      total_rows = @bq_writer.query("SELECT count(*) FROM [#{dataset}.#{table}]")['rows'].first['f'].first['v']
+      @bq_query ||= @bq_writer.query("SELECT * FROM [#{dataset}.#{table}] WHERE RAND() < #{qty * 2}/#{total_rows} LIMIT #{qty}")
+
+      @bq_query['schema']['fields'].each do |f|
+        @schema << f['name']
+      end
+      @schema.uniq!
+      build_records(@bq_query['rows'])
+    end
+
+    def query_db_records(klass, date_attr, qty, start_date, end_date)
+      @db_ids = []
+      db_records =
+        if qty
+          klass.where("#{date_attr} > ? AND #{date_attr} < ?", start_date, end_date).shuffle.first(qty)
+        else
+          klass.where("#{date_attr} > ? AND #{date_attr} < ?", start_date, end_date)
+        end
+      db_records.each { |r| @db_ids << r.id }
+    end
+
+    def collect_bq_records(klass, dataset, table, ids)
+      @bq_query = @bq_writer.query("SELECT * FROM [#{dataset}.#{table}] WHERE #{klass}_record_id IN (#{ids.to_s.gsub(/\[|\]/, '')})")
+    end
+
+    def build_records(rows)
+      @records = []
+      rows.each do |row|
+        values = []
+        row['f'].each_with_index do |value, index|
+          values << [@schema[index], value['v']]
+        end
+        @records << values.to_h
+      end
+    end
+
+    def compare_(records, class_name)
+      @results = []
+      records.each do |record|
+        @fails = []
+        if record['friendly_id'].nil?
+          @results << "Order Record #{record['order_record_id']} Failed, friendly_id is nil!"
```

Comment:
This is code specific to CatPrint, shouldn't really be in a gem...maybe move the actual checks to CatPrint but just provide the ability to do it in here?